build-release.sh

```diff
@@ -1,87 +1,102 @@
-#!/usr/bin/env bash
+#!/bin/bash
 
 set -eEuo pipefail
 
 eval "$(go env)"
 
-# To override the latest git tag as the version, pass something else as the first arg.
-VERSION=${1:-$(git describe --tags --always --dirty)}
+OUT_DIR="${1-dist}"
+# To override the latest git tag as the version, pass something else as the second arg.
+VERSION=${2:-$(git describe --tags --always --dirty)}
 
-# To overwrite the version details, pass something as the second arg. Empty string disables it.
-VERSION_DETAILS=${2-"$(date -u +"%FT%T%z")/$(git describe --always --long --dirty)"}
+# To overwrite the version details, pass something as the third arg. Empty string disables it.
+VERSION_DETAILS=${3-"$(date -u +"%FT%T%z")/$(git describe --always --long --dirty)"}
 
-make_archive() {
-	local FMT="$1" DIR="$2"
+build() {
+    local ALIAS="$1" SUFFIX="${2}"  # Any other arguments are passed to the go build command as env vars
+    local NAME="k6-${VERSION}-${ALIAS}"
 
-	case $FMT in
-	zip)
-		zip -rq9 "$DIR.zip" "$DIR"
-		;;
-	tgz)
-		tar -zcf "$DIR.tar.gz" "$DIR"
-		;;
-	esac
-}
-
-build_dist() {
-	local ALIAS="$1" FMT="${2}" SUFFIX="${3}"  # Any other arguments are passed to the go build command as env vars
-	local DIR="k6-${VERSION}-${ALIAS}"
-
-	local BUILD_ENV=("${@:4}")
-	local BUILD_ARGS=(-o "dist/$DIR/k6${SUFFIX}" -trimpath)
-
-	if [ -n "$VERSION_DETAILS" ]; then
-		BUILD_ARGS+=(-ldflags "-X github.com/loadimpact/k6/lib/consts.VersionDetails=$VERSION_DETAILS")
-	fi
-
-	echo "- Building platform: ${ALIAS} (" "${BUILD_ENV[@]}" "go build" "${BUILD_ARGS[@]}" ")"
+    local BUILD_ENV=("${@:3}")
+    local BUILD_ARGS=(-o "${OUT_DIR}/${NAME}/k6${SUFFIX}" -trimpath)
 
-	# Clean out any old remnants of failed builds.
-	rm -rf "dist/$DIR"
-	mkdir -p "dist/$DIR"
+    if [ -n "$VERSION_DETAILS" ]; then
+        BUILD_ARGS+=(-ldflags "-X github.com/loadimpact/k6/lib/consts.VersionDetails=$VERSION_DETAILS")
+    fi
 
-	# Subshell to not mess with the current env vars or CWD
-	(
-		export "${BUILD_ENV[@]}"
+    echo "- Building platform: ${ALIAS} (" "${BUILD_ENV[@]}" "go build" "${BUILD_ARGS[@]}" ")"
 
-		# Build a binary
-	 	go build "${BUILD_ARGS[@]}"
+    mkdir -p "${OUT_DIR}/${NAME}"
 
-		# Archive it all, native format depends on the platform.
-		cd dist
-		make_archive "$FMT" "$DIR"
-	)
+    # Subshell to not mess with the current env vars or CWD
+    (
+        export "${BUILD_ENV[@]}"
+        # Build a binary
+         go build "${BUILD_ARGS[@]}"
+    )
+}
 
-	# Delete the source files.
-	rm -rf "dist/$DIR"
+package() {
+    local ALIAS="$1" FMT="$2"
+    local NAME="k6-${VERSION}-${ALIAS}"
+    echo "- Creating ${NAME}.${FMT} package..."
+    case $FMT in
+    deb|rpm)
+        # The go-bin-* tools expect the binary in /tmp/
+        [ ! -r /tmp/k6 ] && cp "dist/${NAME}/k6" /tmp/k6
+        "go-bin-${FMT}" generate --file "packaging/${FMT}.json" -a amd64 \
+            --version "${VERSION#v}" -o "dist/k6-${VERSION}-amd64.${FMT}"
+        ;;
```

Comment:
Not certain about that being here (especially without any checks if this will pass). Previously we only needed go to build the binaries for all OSes ... now this will break half way through and also we will upload the deb and rpm files in the release? 

I am not saying that the deb and rpm shouldn't be in the release ... but it seems not all that useful, while being able to build the binaries for everything only running one command was great ;) 

Maybe ... Have a second shell script for building the rpm/debs? And it can error out if `dist/${NAME}/k6` doesn't exist and tell you to run the other one ? or just build it on its own. I am not certain

Reply:
There's already a second shell script for building the rpm/deb: [`gen-packages.sh`](https://github.com/loadimpact/k6/blob/f85389f4efc04cfe0e8c5840e28a7dfc8c595b8b/packaging/gen-packages.sh). The reason I moved it here is because this script already packages the binaries in .zip and .tar.gz, so it makes sense to keep this contained. I'd also like to get rid of the `packaging/` directory entirely, though now that we're back to using the Go tools I guess that won't be happening, and there are also some Windows package assets there that we can't remove.

> now this will break half way through

Why would it break half way through? We already rely on this directory structure for the other packages and there's no error checking there... If anything, I missed using `$OUT_DIR` instead of `dist` in the `cp` command here, so I'll change that.

As for uploading the deb/rpm to the GitHub release, yeah, that might not be needed. I'd argue it would still be useful for anyone who prefers not using the Bintray repositories, and there's no harm in giving more options. But if you feel strongly about it I could filter them out when creating the GH release.